### PR TITLE
mypy --disallow-incomplete-defs

### PR DESCRIPTION
**Pull request by crusaderky**
Add flags to mypy:

allow_incomplete_defs = false
allow_untyped_decorators = false
warn_unreachable = true

### CODE REVIEW

File: distributed/diagnostics/memory_sampler.py [hunk 1]
```diff
@@ -160,6 +164,10 @@ def plot(self, *, align: bool = False, **kwargs):
             See :meth:`~distributed.diagnostics.MemorySampler.to_pandas`
         kwargs
             Passed verbatim to :meth:`pandas.DataFrame.plot`
+
+        Returns
+        =======
+        Output of :meth:`pandas.DataFrame.plot`
```
Comment (crusaderky) [inline distributed/diagnostics/memory_sampler.py:170]: not necessarily a matplotlib object; it depends on the backend

**Comment by fjetter** [comment]
Copying a comment from https://github.com/dask/distributed/pull/6585

I would suggest to enable this file by file (or by module) to allow for faster merging, see https://mypy.readthedocs.io/en/stable/config_file.html#examples

**Comment by fjetter** [comment]
Wouldn't a `warn_return_any` be sufficient to address my concerns in https://github.com/dask/distributed/pull/6585#discussion_r902456095?

**Comment by crusaderky** [comment]
> Wouldn't a `warn_return_any` be sufficient to address my concerns in [#6585 (comment)](https://github.com/dask/distributed/pull/6585#discussion_r902456095)?

No, that means that if f is annotated and it calls g which is not annotated or generic, f can't return what g returned to it without either asserting or casting. e.g. this would fail:
```python
def timeout() -> float:
    return dask.config.get("distributed.something.timeout")
```

**Comment by crusaderky** [comment]
> Copying a comment from #6585
> 
> I would suggest to enable this file by file (or by module) to allow for faster merging, see https://mypy.readthedocs.io/en/stable/config_file.html#examples

That's a good idea; however I spent time on this and, even after blacklisting the biggest offenders, I could not get it below 54 errors in 22 files. It's death by a thousand cuts and I really don't have the time to go through all of them.

**Pull request merged**